cpp/include/cudf/round_robin.hpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2018-2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <vector>
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief Round-robin partition.
+ * 
+ * Returns a new table with rows re-arranged into partition groups and
+ * a vector of row offsets to the start of each partition in the output table.
+ * Rows are assigned partitions based on their row index in the table,
+ * based on the following formula:
+ * partition = (row_index + start_partition) % num_partitions
+ 
+ * Merges two sorted tables into one sorted table
+ * containing data from both tables.
+ *
+ * Example 1:
+ * input:
+ * table => col 1 {0, ..., 12}
+ * num_partitions = 3
+ * start_partition = 0
+ * 
+ * output: pair<table, partition_offsets>
+ * table => col 1 {0,3,6,9,12,1,4,7,10,2,5,8,11}
+ * partition_offsets => {0,5,9}
+ *
+ * Example 2:
+ * input:
+ * table => col 1 {0, ..., 12}
+ * num_partitions = 3
+ * start_partition = 1
+ * 
+ * output: pair<table, partition_offsets>
+ * table => col 1 {0,3,6,9,12,1,4,7,10,2,5,8,11}
+ * partition_offsets => {9,0,5}
```

Comment:
As a user of this function, I'd be a bit surprised that the offsets are not increasing. Since I imagine the use case from @revans2 is to pass the output of this function into `split`, and I _believe_ that function requires the splits to be  increasing, these would be incompatible.

Reply:
@nvdbaranec can you confirm whether or not `contiguous_split/split` requires the split offsets to be increasing?

Reply:
Take `nrows` = 11, `npartitions` = 3
If you round-robin [0,1,2,3,4,5,6,7,8,9,10]

You get: 
partition 0: 0 3 6 9
partition 1: 1 4 7 10
partition 2: 2 5 8

Hence: [[0,3,6,9],[1,4,7,10],[2,5,8]] (round-robin-ed (rr-ed) rows = [0,3,6,9,1,4,7,10,2,5,8])
---------0-1-2-3-4-5 6-7--8 9 10

So, for `start_partition` = 0, this means partition offsets: [0,4,8], in the rr-ed row index re-arrangement.

But, if you want a `start_partition` = 1 then (as I understand it) you want to put whatever was in partition 0 to now be in partition 1, etc.

Hence:
[[2,5,8], [0,3,6,9], [1,4,7,10]] => [8,0,4]
I.e., partition 1 is [0,3,6,9], starting at rr-ed row index 0;
partition 2 is [1,4,7,10], starting at rr-ed row index 4;
and therefore 
partition 0 is [2,5,8] starting now at rr-ed row index 8;

Essentially [0,4,8] gets "rotated" right by `start_partition` positions. Is that not the desired functionality?

Reply:
My untrained eye assumed "start_position" would be an offset of the input, not the output offsets. Seems like the output offsets could be shifted very easily after-the-fact...
```
Example 2:
 * input:
 * table => col 1 {0, ..., 12}
 * num_partitions = 3
 * start_partition = 1
 * 
 * output: pair<table, partition_offsets>
 * table => col 1 {1,4,7,10,0,2,5,8,11,3,6,9,12}
 * partition_offsets => {0,5,9}
```

Reply:
To answer Jake's question - split() (and contiguous_split() expect) the offsets to be increasing. 

```
 * For all `i` it is expected `splits[i] <= splits[i+1] <= input.size()`
```

Similarly, slice requires the same within each pair of slice indices.

Reply:
In that case..."Houston, we've got a problem...". I'm waiting for some feedback from @jlowe on the correct interpretation of `start_partition` or what we can compromise here, in case my interpretation above was correct (one possible compromise might be to forego `start_partition` argument and always return the offsets for `start_partition = 0`).

Reply:
> 
> 
> My untrained eye assumed "start_position" would be an offset of the input, not the output offsets. Seems like the output offsets could be shifted very easily after-the-fact...
> 
> ```
> Example 2:
>  * input:
>  * table => col 1 {0, ..., 12}
>  * num_partitions = 3
>  * start_partition = 1
>  * 
>  * output: pair<table, partition_offsets>
>  * table => col 1 {1,4,7,10,0,2,5,8,11,3,6,9,12}
>  * partition_offsets => {0,5,9}
> ```

@cwharris, so then in the interpretation above `start_partition` would actually mean "the row index from which the round-robin starts", correct?

Reply:
Yes, @jrhemstad is correct that the intent is to call `contiguous_split` after partitioning.  Therefore we need the generated partition offsets to be increasing.

As for the semantics for the `start_partition` argument, it controls which partition we start populating with row values in the round-robin algorithm.  Using the same example above:

Take nrows = 11, npartitions = 3
If you round-robin [0,1,2,3,4,5,6,7,8,9,10] with start_partition=0 then the resulting table arrangement needs to be:
[0,3,6,9,1,4,7,10,2,5,8]
and the partition offsets vector is:
[0,4,8]

If the start_partition changes to 1 then whatever used to be in partition 0 is now in partition 1, partition 1 moves to 2, etc.  It is like a rotate right, but the rotation occurs in the table arrangement rather than the offsets vector.  The output needs to have a table arrangement of:
[2,5,8,0,3,6,9,1,4,7,10]
and the partition offsets vector is:
[0,3,7]

Reply:
@aschaffer I just realized I mistook "start_partition" for "start_position".

But yes, I believe you understand what I meant. I expected `start_position`, so that the first element in the output would be the `start_position`'th element in the input, but the offsets of the partitions would be increasing.

Reply:
Ok, fixed the implementation and tests, accordingly.

Reply:
@cwharris, yes, for a while I was confused too about whether `start_partition` meant a starting row. In fact I prototyped an implementation in Haskell with varying starting row. I actually hoped you're right 'cause the change would be minimal. Unfortunately, that was not the meaning, and the actual meaning complicated a bit the implementation I had. Should be fine now, though. 

I think the lesson to be learnt is that implementation details and examples should be discussed and posted on the feature / issue github page, _before_ the work starts.

Reply:
My apologies for not making this more clear in the issue I filed.  I'll try to provide concrete examples of input and output to help clarify in the future.

Reply:
@jlowe, no worries. I think we're good now. I'll add your comment above (quoted below) to the github issue page: https://github.com/rapidsai/cudf/issues/3648.

Thanks. 

> 
> 
> As for the semantics for the `start_partition` argument, it controls which partition we start populating with row values in the round-robin algorithm. Using the same example above:
> 
> Take nrows = 11, npartitions = 3
> If you round-robin [0,1,2,3,4,5,6,7,8,9,10] with start_partition=0 then the resulting table arrangement needs to be:
> [0,3,6,9,1,4,7,10,2,5,8]
> and the partition offsets vector is:
> [0,4,8]
> 
> If the start_partition changes to 1 then whatever used to be in partition 0 is now in partition 1, partition 1 moves to 2, etc. It is like a rotate right, but the rotation occurs in the table arrangement rather than the offsets vector. The output needs to have a table arrangement of:
> [2,5,8,0,3,6,9,1,4,7,10]
> and the partition offsets vector is:
> [0,3,7]